asiana21-DOCSUP-797 (#117)
BayoNet/ClickHouse: docs/en/sql-reference/functions/other-functions.md

Takes state of aggregate function. Returns result of aggregation (finalized state).

## runningAccumulate {#runningaccumulate}

Accumulates states of an aggregate function for each row of block.  The function resets the state for each new block.

**Syntax**

```sql
runningAccumulate(agg_state[, grouping]);
```

**Parameters**

- `agg_state` — State of the aggregate function. [AggregateFunction](../../sql-reference/data-types/aggregatefunction.md#data-type-aggregatefunction).
- `grouping` — Grouping key. Optional. The state of the function is reset if the `grouping` value is changed. It can be any of the [supported data types](../../sql-reference/data-types/index.md) for which the equality operator is defined.

**Returned value**

- Each resulting row contains a result of the aggregate function, accumulated for all the input rows from 0 to the current position. `runningAccumulate` resets states for each new data block or when the `grouping` value changes.

Type depends on the aggregate function used.

**Examples**

> Let's consider using `runningAccumulate` to find the cumulative sum of numbers with and without grouping.

Reviewer: ```suggestion
Consider how you can use `runningAccumulate` to find the cumulative sum of numbers without and with grouping.
```

Query:

```sql
SELECT k, runningAccumulate(sum_k) as res FROM (SELECT number as k, sumState(k) AS sum_k FROM numbers(10) GROUP BY k ORDER BY k);
```

Result:

```text
┌─k─┬─res─┐
│ 0 │   0 │
│ 1 │   1 │
│ 2 │   3 │
│ 3 │   6 │
│ 4 │  10 │
│ 5 │  15 │
│ 6 │  21 │
│ 7 │  28 │
│ 8 │  36 │
│ 9 │  45 │
└───┴─────┘
```

The subquery generates `sumState` for every number from `0` to `9`. `sumState` returns the state of the [sum](../../sql-reference/aggregate-functions/reference.md#agg_function-sum) function that contains the sum of a single number.

The whole query does the following:

1. For the first row, `runningAccumulate` takes `sumState(0)` and returns `0`. 
2. For the second row, the function merges `sumState(0)` and `sumState(1)` resulting in `sumState(0 + 1)`, and returns `1` as a result.
3. For the third row, the function merges `sumState(0 + 1)` and `sumState(2)` resulting in `sumState(0 + 1 + 2)`, and returns `3` as a result.
4. The actions are repeated until the block ends.

> The following example shows the `groupping` parameter usage:

Reviewer: ```suggestion

The following example shows the `groupping` parameter usage:
```

Takes state of aggregate function. Returns result of aggregation (finalized state).

## runningAccumulate {#runningaccumulate}

> Takes states of the aggregate function, and for each row of block returns the result of aggregate function on merge of states of all previous rows and the current row.

Reviewer: ```suggestion
Accumulates states of an aggregate function for each row of block.  The function reset the state for each new block.
```

**Syntax**

```sql
runningAccumulate(agg_state[, grouping]);
```

**Parameters**

- `agg_state` — State of the aggregate function. [AggregateFunction](../../sql-reference/data-types/aggregatefunction.md#data-type-aggregatefunction).

> - `grouping` — Grouping key. Optional. The state is reset to zero if the parameter value is changed. It can be any of the [supported data types](../../sql-reference/data-types/index.md) for which the equality operator is defined.

Reviewer: ```suggestion
- `grouping` — Grouping key. Optional. The state of the function is reset if the `grouping` value is changed. It can be any of the [supported data types](../../sql-reference/data-types/index.md) for which the equality operator is defined.
```

**Returned value**

> - A column with values that result from accumulating input states for a set of block lines, from the first to the current line. Result of function depends on partition of data to blocks and on order of data in block.

Reviewer: - Each resulting row contains a result of the aggregate function, accumulated for all the input rows from 0 to the current position. `runningAccumulate` resets states for each new data block or at the `grouping` value changes.
```suggestion
- Each resulting row contains a result of the aggregate function, accumulated for all the input rows from 0 to the current position. `runningAccumulate` resets states for each new data block or when the `grouping` value changes.
```

Type depends on the aggregate function used.

**Examples**

Let's consider using `runningAccumulate` to find the cumulative sum of numbers with and without grouping.

Query:

```sql
SELECT k, runningAccumulate(sum_k) FROM (SELECT number as k, sumState(k) AS sum_k FROM numbers(10) GROUP BY k ORDER BY k);
```

Result:

```text
┌─k─┬─runningAccumulate(sum_k)─┐
│ 0 │                        0 │
│ 1 │                        1 │
│ 2 │                        3 │
│ 3 │                        6 │
│ 4 │                       10 │
│ 5 │                       15 │
│ 6 │                       21 │
│ 7 │                       28 │
│ 8 │                       36 │
│ 9 │                       45 │
└───┴──────────────────────────┘
```

> In this case, we have a subquery that generates `sumState` for every number from `0` to `9`. `sumState` is the state of the [sum](../../sql-reference/aggregate-functions/reference.md#agg_function-sum) function that contains the sum of a single number. Let's describe the algorithm of the function.

Reviewer: ```suggestion
The subquery generates `sumState` for every number from `0` to `9`. `sumState` returns the state of the [sum](../../sql-reference/aggregate-functions/reference.md#agg_function-sum) function that contains the sum of a single number.

The whole query does the following:
```

> 1. It takes `sumState(0)` for the first row, and gets the result - `0` from it. 

Reviewer: ```suggestion
1. For the first row, `runningAccumulate` takes `sumState(0)` and returns `0`. 
```

> 2. For the second row, the function combines `sumState(0)` and `sumState(1)`, and gets `sumState(0 + 1)` with the value `1` as result. 
> 3. For the third row, it takes the previous `sumState(0 + 1)`, adds `sumState(2)`, gets `sumState(0 + 1 + 2)`, writes the result `3`.
> 4. The actions are repeated until the block ends.

Reviewer: ```suggestion
2. For the second row, the function merges `sumState(0)` and `sumState(1)` resulting in `sumState(0 + 1)`, and returns `1` as a result.
3. For the third row, the function merges `sumState(0 + 1)` and `sumState(2)` resulting in `sumState(0 + 1 + 2)`, and returns `3` as a result.
4. The actions are repeated until the block ends.
```

> 

Reviewer: ```suggestion
The following example shows the `groupping` parameter usage:

```

Query:

```sql

> SELECT
>     grouping,
>     item,
>     runningAccumulate(state, grouping)
> FROM
> (
>     SELECT
>         number % 6 AS grouping,
>         number AS item,
>         sumState(number) AS state
>     FROM numbers(15)
>     GROUP BY
>         grouping,
>         item
>     ORDER BY
>         grouping ASC,
>         item ASC
> );

Reviewer: ```suggestion
SELECT 
    grouping,
    item,
    runningAccumulate(state, grouping) AS res
FROM 
(
    SELECT 
        toInt8(number / 4) AS grouping,
        number AS item,
        sumState(number) AS state
    FROM numbers(15)
    GROUP BY item
    ORDER BY item ASC
);
```

```

Result:

```text

> ┌─grouping─┬─item─┬─runningAccumulate(state, grouping)─┐
> │        0 │    0 │                                  0 │
> │        0 │    6 │                                  6 │
> │        0 │   12 │                                 18 │
> │        1 │    1 │                                  1 │
> │        1 │    7 │                                  8 │
> │        1 │   13 │                                 21 │
> │        2 │    2 │                                  2 │
> │        2 │    8 │                                 10 │
> │        2 │   14 │                                 24 │
> │        3 │    3 │                                  3 │
> │        3 │    9 │                                 12 │
> │        4 │    4 │                                  4 │
> │        4 │   10 │                                 14 │
> │        5 │    5 │                                  5 │
> │        5 │   11 │                                 16 │
> └──────────┴──────┴────────────────────────────────────┘
> 

Reviewer: ```suggestion
┌─grouping─┬─item─┬─res─┐
│        0 │    0 │   0 │
│        0 │    1 │   1 │
│        0 │    2 │   3 │
│        0 │    3 │   6 │
│        1 │    4 │   4 │
│        1 │    5 │   9 │
│        1 │    6 │  15 │
│        1 │    7 │  22 │
│        2 │    8 │   8 │
│        2 │    9 │  17 │
│        2 │   10 │  27 │
│        2 │   11 │  38 │
│        3 │   12 │  12 │
│        3 │   13 │  25 │
│        3 │   14 │  39 │
└──────────┴──────┴─────┘
```

```

> In this case, the cumulative sum is calculated by group. `sumState` is reset to zero if `grouping` is changed. 

Reviewer: ```suggestion
As you can see, `runningAccumulate` merges states for each group of rows separately.
```

## runningAccumulate {#runningaccumulate}

> Accumulates states of an aggregate function for each row of block.  The function resets the state for each new block.

Reviewer: ```suggestion
Accumulates states of an aggregate function for each row of a data block.  The function resets the state for each new block.
```

Query:

```sql

> SELECT k, runningAccumulate(sum_k) as res FROM (SELECT number as k, sumState(k) AS sum_k FROM numbers(10) GROUP BY k ORDER BY k);

Reviewer: ```suggestion
SELECT k, runningAccumulate(sum_k) AS res FROM (SELECT number as k, sumState(k) AS sum_k FROM numbers(10) GROUP BY k ORDER BY k);
```
